pandas/core/internals/managers.py

```diff
@@ -1945,11 +1945,22 @@ def _compare_or_regex_search(a, b, regex=False):
     is_a_array = isinstance(a, np.ndarray)
     is_b_array = isinstance(b, np.ndarray)
 
+    # GH#32621 replace all pd.NAs to avoid failure of element-wise comparison
```

Comment:
I'm surprised to see this in internals/managers.py. I think if it's only affecting StringArray, the fix should be targeted at na array.

Is fixing this in StringArray._create_arithmetic_method an option?

Reply:
Doesn't quite look like that'll work unfortunately.

It may be worth tracking down exactly where in `.replace` we convert to an ndarray.

Reply:
Hi @TomAugspurger,

It all comes down to comparing with `pd.NA`.  The issue can reproduced with other types of array, not only StringArray. See example below:
```python
import pandas as pd
import numpy as np

rep = {0:1, 1:2}
a = pd.Series([0, 1, pd.NA])
a.replace(to_replace=rep)
# TypeError: Cannot compare types 'ndarray(dtype=object)' and 'int'
```
Given that, it may need to be resolved in internals/managers.py, but not completely sure. Please have a look into my most recent changes.